### ISSUE TITLE
DMs from some people make them disappear from the DM list, and not notified

### ISSUE THREAD

**Issue by Ppjet6** [mattermost/mattermost-server#7769]
#### Summary

Somebody sends you a Direct Message, you see the notification and it instantly disappears, also removing that person from the opened Direct Messages list.
The messages themselves are not gone, but you are not aware of them until you manually look for the person and add it in your Direct Messages list.

It happens only with some people, not everybody, it happens often enough to be noticeable, (people pinging you OOB, "Have you seen my message?" "what message?"), but I haven't managed to reproduce this systematically.

This does not happen when you are already focusing the direct message window.

We also have matterircd running, and a small part is using it, but it can happen to anybody from what I've seen (both not using it, one using it not the other, etc.).

This is obviously a deal breaker if we can't reliably contact people via an IM application.

I also reported this to matterircd in case this is linked, 42wim/matterircd#131.

#### Steps to reproduce

I don't know. Please help me debug this.

Mattermost version: 4.2
Matterircd version: 0.15.0

#### Expected behavior

- Get notified when somebody sends you a direct message
- Have that person appear in the direct messages

#### Observed behavior (that appears unintentional) 

- Notification instantly disappear
- People who sent messages don't stay in the direct messages list

**Comment by lindy65** [comment]
Hi @Ppjet6,

Thank you for your report,

Could we have your help updating to the [latest version](https://about.mattermost.com/download/) of Mattermost to see if the issue still persists?

Let us know how it goes!

**Comment by lindy65** [comment]
Hi @Ppjet6 

Have you had a chance to update to the latest version of Mattermost yet?  You might like to wait until tomorrow (11/16) when the latest version (4.4) is released.

Please let us know if your issue is resolved after updating :)

**Comment by lindy65** [comment]
Hi @Ppjet6, we haven't received an update so we'll assume that the problem is fixed or is no longer valid. 

If you still experience the same problem, try upgrading to the [latest version](https://about.mattermost.com/download/). 

If the issue persists, reopen this issue with the relevant information and we'd be glad to help you where we can.

**Comment by jasonblais** [comment]
@Ppjet6 How frequently do you notice this? Also, confirming this doesn't just happen with messages made via the matterircd bridge?

/cc @amyblais

**Comment by Ppjet6** [comment]
I just tested being disconnected from the bridge entirely. I had also disconnected every client, and connected from a new profile on Firefox.

It happens reliably every time from a small fraction of users, always the same users. Once it happened for a user I know it will always happen.

**Comment by amyblais** [comment]
@Ppjet6 Thank you for reporting this,

Can you help with more details on the following:

1. what browser are you using?
2. can you gather any details about the users that this happens with, e.g. what browser they are using, are they all using v4.8, etc. (just want to get a sense of if there is a pattern)?
3. would you be able to gather logs from around one of the times the issue occurs?

**Comment by Ppjet6** [comment]
Here are a few logs from mattermost.
[mattermost-7769.log](https://github.com/mattermost/mattermost-server/files/1971048/mattermost-7769.log)

I have asked userX, whose messages I am never notified for, to send me multiple messages via the web app.

I really wish the websocket logs included a user id for `user_typing`, because even if the timing corresponds more or less, I can't be sure it's userX over an instance with 100+ people. (I did try to choose a time slot with the less online users possible).

Logs from matterircd do not give more context for this unfortunately, (and a bit more annoying to clean, which I why I won't paste them). I don't see any message coming through, I didn't even see this `user_typing` message.

**Comment by amyblais** [comment]
Hi @Ppjet6 Following up on this - does @icelander's suggestions sound good and do you need any assistance? Please let us know and we are happy to help / investigate this issue further!

**Comment by amyblais** [comment]
Hi @Ppjet6, we haven't received an update so we'll assume that the problem is fixed or is no longer valid. 

If you still experience the same problem, try upgrading to the [latest version](https://about.mattermost.com/download/). 

If the issue persists, reopen this issue with the relevant information and we'd be glad to help you where we can.

**Comment by tony1432** [comment]
@Ppjet6 I've seen this issue as well. Do you per chance have Data Retention enabled? 
I'm yet to reproduce it, but I think it might occur when the direct message channel has been made empty by Data Retention and a new message is added.

**Comment by Ppjet6** [comment]
@tony1432, we are not using EE, we don't have that feature!

**Comment by lindy65** [comment]
Hi @Ppjet6 @tony1432 @amyblais 

I've done some testing in collaboration with one of our engineers to try and reproduce

> Somebody sends you a Direct Message, you see the notification and it instantly disappears, also removing that person from the opened Direct Messages list.

I tried reproducing by having the person already listed in direct messages, listed in direct messages and closing the direct message with them, as well as not having received a direct message from the person ever before.  

Because `tony1432` mentioned this might have something to do with data retention (although you've been seeing this since v4.2 which was before the feature was shipped) I tried reproducing the issue before and after running a data retention job but still wasn't able to reproduce.

Our engineer has asked for the javascript console log errors you get when this issue occurs.  Can you help by adding a screenshot to this issue the next time this happens please?

We'll be happy to continue troubleshooting once we have further information from you.  Thanks!

**Comment by Ppjet6** [comment]
Still on 4.8, sorry, we're planning to upgrade soon.

I've asked a user with whom I've got the issue to send me a message, they don't see anything in the console when they do this (using chromium. I was physically besides them), they do see POST requests to `/api/v4/posts`.
On my end, I see literally no activity in the console or network tabs, apart from the regular pings to `/api/v4/users/status/ids` (firefox nightly _and_ chromium latest, disconnected from the bridge).

After all this, if I go and manually open a direct message tab, I do see the post history.

Maybe of interest, this direct message tab doesn't stay in the list, unlike others. If I close my browser and reopen it, that tab is not in the list anymore. I can reproduce all the time.

Also, when I open that direct message tab, it seems I get `404` on `/api/v4/channels/ishfd3uni3yt98pu4xix4w6kur/members/me`. I can reproduce with everybody I've had issues with, and this doesn't happen with others.

```json
{"id":"store.sql_channel.get_member.missing.app_error","message":"No channel member found for that user ID and channel ID","detailed_error":"channel_id=ishfd3uni3yt98pu4xix4w6kuruser_id=z7m68jcp37bgb8mjcbnauz5eww,sql: no rows in result set","request_id":"ed68zx4ujtr98dmbayoipmcdmr","status_code":404}
```

Seems that the `channel_id` field and `user_id` field are not separated in that `detailed_error` field, not sure if that's on purpose.

Another note, I haven't heard anybody complain the other way yet. (me sending messages and them not being notified). Maybe it is also just that they haven't noticed, and I don't initiate private messages that often either.

**Comment by lindy65** [comment]
Thanks for the feedback @Ppjet6,

Do you have `CloseUnusedDirectMessages` setting enabled in your config.json?  This was introduced in v4.5 and just wondered whether it could be causing issues for you?

The screen for the user looks like this, accessible from Main Menu > Account Settings > Sidebar > Sidebar Settings >Automatically close direct messages

![image](https://user-images.githubusercontent.com/13020833/41367356-11ed6a26-6f3f-11e8-86dc-6d41dbffb117.png)

**Comment by Ppjet6** [comment]
I changed to `Never`, will keep you updated.

**Comment by Ppjet6** [comment]
No visible improvement with the change.

I did in fact gain back old direct message tabs, that I manually cleaned, but none of these were from contacts with whom I have these issues.

**Comment by lindy65** [comment]
Thanks @Ppjet6,

Can you share what the users you have this issue with have the above setting set to?

**Comment by Ppjet6** [comment]
We updated to 5.0 last week, and I still observe the issue with the same set of people. Actually a new person entered that set yesterday.. this is getting annoying.

@lindy65, that last person has the setting set to `Never` apparently.

Also an observation, messages don't seem to arrive to me straight away, there is a delay between when they send the messages and when I receive them. I can see that due to the fact that the conversation is out of sequence.

**Comment by amyblais** [comment]
@Ppjet6 - giving you an update that I will gather all your information and think about the next step in troubleshooting this issue! Thank you for your patience.

**Comment by amyblais** [comment]
@Ppjet6 Do you know if the users you are seeing this issue with are system admins / members / a mix of both?

I'll also ask an engineer to help take a look at the 404 error message you posted earlier to see if they have ideas.

**Comment by amyblais** [comment]
@Ppjet6 Update: as a next step, I posted about this for engineers and community members on our [build server](https://pre-release.mattermost.com/signup_user_complete/?id=f1924a8db44ff3bb41c96424cdc20676) in [Peer-to-peer Help channel](https://pre-release.mattermost.com/core/channels/peer-to-peer-help) to see if anyone has ideas on the 404 error message you saw earlier.

**Comment by amyblais** [comment]
@Ppjet6 Another suggestion would be that you could test out with the affected users if you see the same issue on our build server that I linked in my previous comment ^ Everyone is welcome to join and test it out!

**Comment by amyblais** [comment]
Hi @Ppjet6, we haven't received an update for a while so I'll close this Issue for now.

If the problem persists, please reopen this issue with relevant information and we'll continue to troubleshoot this.

**Comment by Ppjet6** [comment]
Good news!

I don't yet have the exact steps to reproduce, but we pinned it down to the following:

- Start 1:1 discussion with someone (from the bridge or MM doesn't matter)
- From MM, send *some kind of message* that the bridge doesn't really understand (need to figure what exactly). At this point the bridge creates a new channel for this
- PART the new channel. That causes MM to leave the 1:1, displaying `You left the channel.` from System, to me and to the other party.

And from that point on it's downhill.

The channel is removed from the list of DMs I have, and if from there I reopen it, it's still in the same state, i.e., with me not in the `channelmembers` table.

If I had myself back to that table manually, and logout/login, it works again.

I think there are multiple issues here.
First the bridge that doesn't understand everything and creates that channel. I need to find when that happens and report.
Then the webui that doesn't rejoin? Maybe it's not excepting a 1:1 Direct Message to be left at all? If so that should certainly be reflected in the API, to prevent it.

Now I will be wandering off figuring out in what channel to add myself again..

Cc @42wim.

**Comment by jasonblais** [comment]
Thanks @Ppjet6! There was another thread with a similar problem and we might have found the cause: https://github.com/mattermost/mattermost-server/issues/8619#issuecomment-410919448

Let me know if that covers it?

**Comment by immolate-d** [comment]
I can confirm this issue is still happening, affects all 200 people in my unit. Our administrator kept telling people last year that disappearing DMs were a known issue. I just saw it today, here is what happened:

1) My phone chime notification goes off
2) I see a mattermost notification on the screen
3) I unlock the screen and open mattermost
4) No no new DM is there. Completely gone
5) I opened a bunch of new DMs with people I didn't have current conversations with. After about 50 tries I found the right person, and there the message was from 10 mins ago.

I'm using Android. Mattermost Server Version 5.27.0 (Build 5.27.0)
App Version: 1.39.0 (Build 3000341)
